txpipe/data_types/types.py

```diff
@@ -469,6 +438,136 @@ def to_ccl(self, **kwargs):
 
         return ccl.Cosmology(**inits)
 
-class QPFile(DataFile):
-    # TODO: Flesh this out
-    suffix = "hdf5"
+
+class QPBaseFile(HDFFile):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.mode == "r":
+            self._metadata = None
+
+    @property
+    def metadata(self):
+        import tables_io
+        if self._metadata is not None:
+            return self._metadata
+        meta = tables_io.io.readHdf5GroupToDict(self.file["meta"])
+        self._metadata = meta
+        return meta
+
+class QPPDFFile(QPBaseFile):
+    def iterate(self, chunk_rows, rank=0, size=1):
+        import qp
```

Comment:
dumb questions, why is this better to import in the functions?

Reply:
You mean inside the function instead of at top level? That's to avoid loading QP until it's actually needed. It takes a while to load as it has so many dependencies. Also I avoid importing stuff until runtime throughout TXPipe for the same reason, and so that the launcher env can be simpler.